### PR TITLE
SONARJAVA-1296 resolve this() and super() constructors

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 1]
```diff
@@ -391,21 +391,29 @@ private Resolution findMethod(Env env, JavaType callSite, JavaType site, String
 
   private Resolution findMethod(Env env, JavaType callSite, JavaType site, String name, List<JavaType> argTypes, List<JavaType> typeParams, boolean autoboxing) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR The Cyclomatic Complexity of this method "findMethod" is 13 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 2]
```diff
@@ -672,7 +680,19 @@ boolean isInheritedIn(JavaSymbol symbol, JavaSymbol.TypeJavaSymbol clazz) {
       case Flags.PUBLIC:
         return true;
       case Flags.PRIVATE:
-        return symbol.owner() == clazz;
+        // FIXME take only constructors into account when all the members should be JLS-6.6.1
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 3]
```diff
@@ -672,7 +680,19 @@ boolean isInheritedIn(JavaSymbol symbol, JavaSymbol.TypeJavaSymbol clazz) {
       case Flags.PUBLIC:
         return true;
       case Flags.PRIVATE:
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Reduce this switch case number of lines from 13 to at most 5, for example by extracting code into methods. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1151)

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 4]
```diff
@@ -672,7 +680,19 @@ boolean isInheritedIn(JavaSymbol symbol, JavaSymbol.TypeJavaSymbol clazz) {
...
+          JavaSymbol topLevel2 = clazz;
+          while (!topLevel1.owner.isPackageSymbol()) {
+            topLevel1 = topLevel1.owner;
+          }
+          while (!topLevel2.owner.isPackageSymbol()) {
+            topLevel2 = topLevel2.owner;
+          }
+          return topLevel1 == topLevel2;
+        }
+        return symbol.owner == clazz;
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Change this comparison to use the equals method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)

File: java-checks/src/main/java/org/sonar/java/checks/CallSuperMethodFromInnerClassCheck.java [hunk 5]
```diff
@@ -79,7 +79,8 @@ public MethodInvocationVisitor(Symbol.TypeSymbol classSymbol) {
     @Override
     public void visitMethodInvocation(MethodInvocationTree tree) {
       Symbol symbol = tree.symbol();
-      if (symbol.isMethodSymbol() && tree.methodSelect().is(Tree.Kind.IDENTIFIER) && isInherited(symbol) && outerClassHasMethodWithSameName(symbol)) {
+      if (symbol.isMethodSymbol() && !"<init>".equals(symbol.name()) && tree.methodSelect().is(Tree.Kind.IDENTIFIER) && isInherited(symbol)
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CallSuperMethodFromInnerClassCheck.java]: MAJOR Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)

File: java-checks/src/main/java/org/sonar/java/checks/InnerStaticClassesCheck.java [hunk 6]
```diff
@@ -123,18 +123,18 @@ private int referenceInstance(Symbol symbol) {
...
   }
 
-  private int fromInstance(Symbol owner) {
+  private int fromInstance(Symbol symbol, Symbol owner) {
     int i = -1;
     Type ownerType = owner.type();
     for (Symbol outerClass : outerClasses) {
       i++;
-      if (owner.equals(outerClass) || (ownerType != null && owner.isTypeSymbol() && outerClass.type().isSubtypeOf(ownerType))) {
+      if (symbol.equals(outerClass) || owner.equals(outerClass) || (ownerType != null && owner.isTypeSymbol() && outerClass.type().isSubtypeOf(ownerType))) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/InnerStaticClassesCheck.java]: MAJOR Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues:
* MAJOR 7 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Reduce the number of returns of this method 6, down to the maximum allowed 5.](https://github.com/SonarSource/sonar-java/blob/98fdb3ad135ad0fc10d639aa7ac1d63a22426903/java-squid/src/main/java/org/sonar/java/resolve/Resolve.java#L678) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1142)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues:
* MAJOR 7 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Reduce the number of returns of this method 6, down to the maximum allowed 5.](https://github.com/SonarSource/sonar-java/blob/98fdb3ad135ad0fc10d639aa7ac1d63a22426903/java-squid/src/main/java/org/sonar/java/resolve/Resolve.java#L678) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1142)

**Comment by m-g-sonar** [comment]
Work on this topic is postponed to version `3.8` of java plugin. Moreover, current solution is not satisfactory.

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues:
* MAJOR 7 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Reduce the number of returns of this method 6, down to the maximum allowed 5.](https://github.com/SonarSource/sonar-java/blob/98fdb3ad135ad0fc10d639aa7ac1d63a22426903/java-squid/src/main/java/org/sonar/java/resolve/Resolve.java#L678) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1142)

**Comment by m-g-sonar** [comment]
will be handled by https://github.com/SonarSource/sonar-java/pull/487

**Pull request closed without merging**